// server/server.js
var http = require('http');
var util = require('util');
var anyBody = require('body/any');

function start() {
    var log4js = require('log4js');
    log4js.configure({ appenders: [ { type: "console", layout: { type: "basic" } } ], replaceConsole: true })

    var fs = require('fs');
    var m_db = require('sql/sqlite/db.js');
    http.createServer(function(request, response) {
        util.log('http server req', request);
        if (response.url === '/favicon.ico') return; /////// i don't want this in the log for now (TODO remove it)
        if ( request.method === 'POST' ) {
            anyBody(request, response, function (err, body) {
                if ( err ) {
                    throw err;
                }
                util.log('http server req body', body);
                request.body = body;
                handle(request, response);
            });
        } else {
            handle(request, response);
        }

        function handle(request, response) {
            var db1 = m_db.db('test.db').open('test.db');
            console.log('http request', request.url);
            var router = require('./router.js');
            router.route(request, function(err, moduleName, controllerName, actionName, query) {
                console.log('routed: ' + request.url, 'to', err, moduleName, controllerName, actionName, query);

                if ( err ) {
                    throw err;
                }

                var controllerFileName = './app/' + moduleName + '/' + controllerName + '.js';
                console.log(controllerFileName);
                var controller = require(controllerFileName);
                if ( !controller ) {
                    throw 'controller file ' + controllerFileName + 'not found';
                }

                var method = controller[actionName];
                
                if ( typeof(method) === 'undefined' ) {
                    throw 'method not found';
                }

                var internalRequest = { 
                    db: db1, 
                    moduleName: moduleName, controllerName: controllerName, actionName: actionName, 
                    orig: request, url: request.url, query: query,
                    body: request.body
                };
                var internalResponse = { 
                    request: internalRequest, 
                    returnCode: 200, 
                    contentType: 'text/html', 
                    text: null, 
                    orig: response 
                };

                method(internalRequest, internalResponse, responseCallback);
            });
        }
    }).listen(8888);
}

function responseCallback(response) {
    response.request.db.close();
    response.orig.writeHead(response.returnCode, {"Content-Type": response.contentType});
    console.log('responseCallback data', response.data);
    if ( response.text === null ) { // TODO arg this is a hack
        var data = response.data; // TODO this should really be cloned...
        data.module = response.request.moduleName;
        data.controller = response.request.controllerName;
        data.action = response.request.actionName;
        response.text = JSON.stringify(data);
    }
    console.log('responseCallback text', response.text);
    response.orig.write(response.text);
    response.orig.end();

}

exports.start = start;
