// web/server/controller/DefaultController.js
/*
 * Copyright (C) 2015-2017 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

const async = require('async');
const _ = require('lodash');
const q = require('q');
const sql_ConditionSet = require('sql/ConditionSet.js');
const bjoo = require('BJOObject.js');

/**
 * @class web.server.DefaultController
 * @construcor
 * @param {Object} config
 */
function DefaultController(config) {
    bjoo.BJOObject.call(this);
    this._config = config;
}

/**
 * @param {Object} config
 */
function create(config) {
    return new DefaultController(config);
}

DefaultController.prototype = new bjoo.BJOObject();
DefaultController.prototype.constructor = DefaultController;
exports.DefaultController = DefaultController;
exports.create = create;

/**
 * The edit action
 * @method edit
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.editAction = function(entitySet, request, response) {
    return entitySet.loadEntityById(request.query.id).then((entity) => {
        return entity.toJson();
    }).then((row) => {
        return _.merge(response, {
            action: 'edit', 
            id: row.id,
            row: row
        });
    });
}

/**
 * The edit action
 * @method list
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.listAction = function(entitySet, request, response) {
    let reqConds = [];
    if ( request.body !== undefined && request.body.conditions !== undefined ) {
        reqConds = request.body.conditions;
    }
    const conditionSet = sql_ConditionSet.createFromWeb(entitySet.getTable(), reqConds);
    return entitySet.findEntities(conditionSet).then((entities) => { // TODO, not ALL
        console.log('listAction entities are', entities);
        return q.all(_.map(entities, (entity) => { 
            return entity.toJson(); 
        }))
    }).then((rows) => {
        console.log('listAction rows are', rows, 'old response is', response);
        response.contentType = 'text/json';
        response.json = {
            rows: rows
        };
        return response;
    });
    // TODO fetch counts, other aggregates?
}

/**
 * The edit action
 * @method saveField
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.saveFieldAction = function(entitySet, request, response) {
    if ( request.body.id !== undefined ) { // we have an id, update the entity
        return entitySet.loadEntityById(request.body.id).then((entity) => {
            entity.getField(request.body.field.name).setValue(request.body.field.value);
            return entity.save().then(() => { return entity; });
        }).then((entity) => {
            return entity.toJson();
        }).then((row) => {
            return _.merge(response, { row: row, id: row.id, performedAction: 'updated' });
        }).catch((err) => {
            return _.merge(response, { performedAction: null, state: 'error', errors: [ err.message ] });
        });
    } else { // insert
        const entity = entitySet.createEntity();
        const row = request.body.row;
        _.each(row, ( field ) => {
            entity.getField(field.name).setValue(field.value); // TODO make convenience @method in Entity for that
        });
        return entity.save()
            .then(() => { return entity.toJson(); })
            .then((row) => {
                return _.merge(response, { row: row, id: row.id, performedAction: 'inserted' });
            }).catch((err) => {
                return _.merge(response, { performedAction: null, state: 'error', errors: [ err.message ] });
            });
    }
}

/**
 * The edit action
 * @method count
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.countAction = function(boSet, request, response) {
    var reqConds = [];
    if ( typeof(request.body) !== 'undefined' ) {
        reqConds = request.body.conditions;
    }
    var conditions = sql_ConditionSet.conditionSetFromWeb(boSet.daoSet().table(), reqConds);
    boSet.countByConditions(conditions, function(err, count) {
        if ( err ) throw err;
        var templateRow = {fields: boSet.templateBo().fieldsToJson() };
        response.data = { count: count, templateRow: templateRow };
        responseCallback(response);
    });
}

/**
 * The edit action
 * @method fetchAggregateRow
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.fetchAggregateRowAction = function(bo, request, response) {
    bo.calculateCalcFields([], function(err) {
        if ( err ) throw err;
        response.data = { row: { fields: bo.fieldsToJson() } };
        responseCallback(response);
    });
}

/**
 * The edit action
 * @method fetchTemplateRow
 * @param {entity.EntitySet} entitySet
 * @param {Object} request
 * @param {Object} response we receive a respnse, which we will then amend
 */
DefaultController.prototype.fetchTemplateRowAction = function(boSet, request, response) {
    response.data = {fields: boSet.templateBo().fieldsToJson() };
    responseCallback(response);
}
