// web/server/controller/IndexController.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

const deps = {
    q: require('q'),
    _: require('lodash'),
    fs: require('fs'),
    bjoo: require('BJOObject.js'),
};

/**
 * @class web.server.controller.IndexController
 * @construcor
 */
function IndexController(config) {
    deps.bjoo.BJOObject.call(this);
    if ( arguments.length > 0 ) {
        this._config = config;
    }
}

function create(config) {
    return new IndexController(config);
}

IndexController.prototype = new deps.bjoo.BJOObject();
IndexController.prototype.constructor = IndexController;
IndexController.prototype._config = {};

exports.IndexController = IndexController;
exports.create = create;

IndexController.prototype.indexAction = function(request, response) {
    const filePath = this._config.web.server.documentRoot + '/html/index.html';

    return deps.q.nfbind(deps.fs.readFile)(filePath, 'utf8').then( (text) => {
        return deps._.merge(response, { text: text });
    });
};

IndexController.prototype.serveFileAction = function(request, response) {
    const filePath = this._config.web.server.documentRoot + '/' + request.url;

    return deps.q.nfbind(deps.fs.readFile)(filePath, 'utf8').then( (text) => {
        return deps._.merge(response, { text: text });
    });
};
