// model/EntityModel.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * a PrimaryEntityModel is a model for an entity which possesses an id field named 'id'
 *
 * an EntityModel is a model for an entity, i.e. something that resides in the database
 * ... that is not necessarily an id-having Entity ( that is a PrimaryEntity)
 *
 * data will reside in the _data property, and Model will create accessors for it
 */

"use strict";

const _ = require('lodash');
const bjoo = require('BJOObject.js');
const sql_Table = require('sql/Table.js');
const sql_Field = require('sql/Field.js');
const sql_ValueField = require('sql/ValueField.js');
const sql_LookupField = require('sql/LookupField.js');
const sql_LookupIdField = require('sql/LookupIdField.js');
const sql_Query = require('sql/Query.js');
const sql_Filter = require('sql/Filter.js');
const Errors = require('Errors.js');

function EntityModel() {
    bjoo.BJOObject.call(this);
}

function create(/* sql_DB.DB */ db, /* sql_Table */ table) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            if ( !(table instanceof sql_Table.Table) ) { throw new Error('table should be of type sql_Table.Table', table); }
        /* @DEVELOPMENT } */

    }
    let model = new EntityModel();
    model.setDb(db);
    model.setTable(table);
    return model;
}

EntityModel.prototype = new bjoo.BJOObject();
EntityModel.prototype.constructor = EntityModel;

exports.EntityModel = EntityModel;
exports.create = create;

EntityModel.prototype.setDb = function(/* sql_DB.DB */ db) {
    this._db = db;
};

EntityModel.prototype.setTable = function(/* sql_Table.Table */ table) {
    this._table = table.cloneWithValueFields();
    if ( this._table.getField('id') === undefined ) {
        this._table.addField(sql_ValueField.create('id', sql_Field.DataType.int, 'ID', null));
    }
};

EntityModel.prototype.setValuesFromRow = function(/* Array */ row) {
    _.each(this._table.getDatabaseFields(), ( field ) => {
        field.setValue(row[field.getName()]);
    });
};

EntityModel.prototype.save = function() {
    if ( this._table.getField('id').getValue() !== null ) {
        console.log('will do update, id value is', this._table.getField('id').getValue());
        return this._update();
    } else {
        return this._insert();
    }
};

EntityModel.prototype._insert = function() {
    return this._db.allSql('SELECT MAX(id)+1 AS id FROM ' + this._table.getName()).then( (rows) => { // TODO this should be done via a query obj
        if ( rows[0]['id'] === null ) {
            this.setId(1);
        } else {
            this.setId(rows[0]['id']);
        }
        const query = sql_Query.insert()
            .into(this._table)
            .setFields(this.getDatabaseFields())
        return this._db.run(query);
    }).catch( ( err ) => {
        console.log('error in _insert', err);
        throw new Error(err);
    });
};

EntityModel.prototype._update = function() {
    const cond = sql_Filter.create()
        .setField(this._table.getField('id'))
        .setOp(sql_Filter.Op.eq)
        .setCompareTo(this._table.getField('id'));
    const query = sql_Query.update()
        .setTable(this._table)
        .setFields(this.getDatabaseFields())
        .where(cond);
    return this._db.run(query);
};


EntityModel.prototype.getDatabaseFields = function() {
    return _.filter(this.getFieldsAsList(), ( field ) => { 
        console.log('field', field, 'instof', (field instanceof sql_LookupField.LookupField));
        return !(field instanceof sql_LookupField.LookupField); 
    });
};

EntityModel.prototype.getId = function() {
    return this._table.getField('id').getValue();
};
EntityModel.prototype.setId = function(id) {
    this._table.getField('id').setValue(id);
    return this;
};

EntityModel.prototype.getFieldsAsList = function() {
    return this._table.getFieldsAsList();
};

EntityModel.prototype.getTable = function() {
    return this._table;
};

EntityModel.prototype.getAttributeNames = function() {
    return _.map(this._table.getFieldsAsList(), ( field ) => { return field.getName(); });
};

EntityModel.prototype.getAttributeValue = function(name) {
    return this._table.getField(name).getValue();
};

EntityModel.prototype.setAttributeValue = function(name, value) {
    this._table.getField(name).setValue(value);
    return this;
};
