// model/EntityModel.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */


"use strict";

const _ = require('lodash');
const q = require('q');
const bjoo = require('BJOObject.js');
const sql_Table = require('sql/Table.js');
const sql_Field = require('sql/Field.js');
const sql_ValueField = require('sql/ValueField.js');
const sql_LookupField = require('sql/LookupField.js');
const sql_Query = require('sql/Query.js');
const sql_Filter = require('sql/Filter.js');
const Errors = require('Errors.js');

/**
 * EntityModel is a model for an entity 
 *
 *
 * - It is assumed that an entity has a field 'id'
 *
 * an EntityModel is a model for an entity, i.e. something that resides in the database
 *
 * @class EntityModel
 * @constructor
 */
function EntityModel() {
    bjoo.BJOObject.call(this);
}

/**
 * External constructor
 *
 * @method create
 * @param {sql_DB.DB} db
 * @param {sql_Table.Table} table
 * @return model_EntityModel.EntityModel
 */
function create(/* sql_DB.DB */ db, /* sql_Table */ table) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            if ( !(table instanceof sql_Table.Table) ) { throw new Error('table should be of type sql_Table.Table', table); }
        /* @DEVELOPMENT } */

    }
    let model = new EntityModel();
    model.setDb(db);
    model.setTable(table);
    return model;
}

EntityModel.prototype = new bjoo.BJOObject();
EntityModel.prototype.constructor = EntityModel;

exports.EntityModel = EntityModel;
exports.create = create;

/**
 * @method setDb
 * @param {sql_DB.DB} db DB
 */
EntityModel.prototype.setDb = function(/* sql_DB.DB */ db) {
    this._db = db;
    return this;
};

/**
 * set the table
 * NOTE: we need to clone the table first
 *
 * @method setTable
 * @param {sql_Table.Table} table
 */
EntityModel.prototype.setTable = function(/* sql_Table.Table */ table) {
    this._table = table.clone();
    if ( this._table.getField('id') === undefined ) {
        this._table.addField(sql_ValueField.create('id', sql_Field.DataType.int, 'ID', null));
    }
    return this;
};

/**
 * @method setValuesFromRow
 * @param {Array} row
 */
EntityModel.prototype.setValuesFromRow = function(/* Array */ row) {
    _.each(this._table.getPersistentFields(), ( field ) => {
        field.setValue(row[field.getName()]);
    });
    return this;
};

/**
 * this is async, because it will use field.getValue()!
 * @method validate
 * @async
 * @return {Promise}
 */
EntityModel.prototype.validate = function() {
    return q.all(_.map(this._table.getPersistentFields(), (field) => { return field.getValue().then((value) => { return field.validate(value, field); })}));
};

/*
 * @async
 * @method save
 * @return {Promise}
 */
EntityModel.prototype.save = function() {
    return this.validate().then( () => {
        return this._table.getField('id').getValue().then( ( id ) => {
            if ( id !== null ) {
                return this._update();
            } else {
                return this._insert();
            }
        });
    });
};

/**
 * @private
 * @method _insert
 * @return {Promise}
 */
EntityModel.prototype._insert = function() {
    return this._db.allSql('SELECT MAX(id)+1 AS id FROM ' + this._table.getName()).then( (rows) => { // TODO this should be done via a query obj
        if ( rows[0]['id'] === null ) {
            this.setId(1);
        } else {
            this.setId(rows[0]['id']);
        }
        const query = sql_Query.insert()
            .into(this._table)
            .setFields(this.getPersistentFields())
        return this._db.run(query);
    }).catch( ( err ) => {
        console.error('error in _insert', err);
    });
};

/**
 * @private
 * @method _update
 * @return {Promise}
 */
EntityModel.prototype._update = function() {
    const cond = sql_Filter.create()
        .setField(this._table.getField('id'))
        .setOp(sql_Filter.Op.eq)
        .setCompareTo(this._table.getField('id').getValue());
    const query = sql_Query.update()
        .setTable(this._table)
        .setFields(this.getPersistentFields())
        .where(cond);
    return this._db.run(query);
};


/**
 * @method getPersistentFields
 * @return {Boolean}
 */
EntityModel.prototype.getPersistentFields = function() {
    return _.filter(this.getFieldsAsList(), ( field ) => { 
        return field.getIsPersistent();
    });
};

/**
 * @method getId
 */
EntityModel.prototype.getId = function() {
    return this._table.getField('id').getValue();
};
/**
 * @method setId
 * @param id
 */
EntityModel.prototype.setId = function(id) {
    this._table.getField('id').setValue(id);
    return this;
};

/**
 * @method getFieldsAsList
 */
EntityModel.prototype.getFieldsAsList = function() {
    return this._table.getFieldsAsList();
};

/**
 * @method getTable
 */
EntityModel.prototype.getTable = function() {
    return this._table;
};

/**
 * @method getAttribute
 * @method getField
 */
EntityModel.prototype.getAttribute = EntityModel.prototype.getField = function(name) {
    return this._table.getField(name);
};
/**
 * @method getAttributeNames
 * @method getFieldNames
 */
EntityModel.prototype.getAttributeNames = EntityModel.prototype.getFieldNames = function() {
    return _.map(this._table.getFieldsAsList(), ( field ) => { return field.getName(); });
};

/**
 * @method getAttributeValue
 * @method getFieldValue
 */
EntityModel.prototype.getAttributeValue = EntityModel.prototype.getFieldValue = function(name) {
    return this._table.getField(name).getValue();
};

/**
 * @method setAttributeValue
 * @param {String} name
 * @param {Object} value
 */
EntityModel.prototype.setAttributeValue = function(name, value) {
    this._table.getField(name).setValue(value);
    return this;
};
