// sql/Aggregate.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

var util = require('util.js');
var bjoo = require('BJOObject.js');

var Type = { sum: 'sum', count: 'count' };

function Aggregate(type, field) {
    /* @DEVELOPMENT */ if ( arguments.length !== 2 ) { throw new Error('Illegal argument length '  + arguments.length); }  
    /* @DEVELOPMENT */ if ( !(util.isString(type)) ) { throw new Error('typecheck failed'); }
    this._type = type;
    this._field = field;
    this._alias = null;
}

Aggregate.prototype = new bjoo.BJOObject();
Aggregate.prototype.constructor = Aggregate;

Aggregate.prototype.type = function(type) {
    if ( typeof(type) !== 'undefined' ) {
        this._type = type;
        return this;
    }
    return this._type;
};

Aggregate.prototype.field = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._field = field;
        return this;
    }
    return this._field;
};

Aggregate.prototype.alias = function(alias) {
    if ( typeof(alias) !== 'undefined' ) {
        this._alias = alias;
        return this;
    }
    return this._alias;
};

function sum(field) { return new Aggregate(Type.sum, field); }
function count(field) { return new Aggregate(Type.count, field); }
function aggregate(type, field) { return new Aggregate(type, field); }

exports.Type = Type;
exports.Aggregate = Aggregate;
exports.aggregate = aggregate;
exports.sum = sum;
exports.count = count;
