// sql/condition.js
/*
 * Copyright (C) 2015 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of BJO2.
 *
 * Mapitor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

var _ = require('underscore');
var m_bjoo = require('BJOObject.js');
var util = require('util.js');
var m_sql_field = require('sql/field.js');

var Op = { eq: 'eq', gt: 'gt', lt: 'lt', gte: 'gte', lte: 'lte', ne: 'ne', like: 'like' };

/**
 * @param {m_sql_field.Field} field
 * @param {Op} op
 * @param {any} compareTo
 */
function Condition(p_field, p_op, p_compareTo) {
    m_bjoo.BJOObject.call(this);
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */ 
            if ( arguments.length !== 3 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + 3); }
            if ( !(p_field instanceof m_sql_field.Field) ) { throw new Error('typecheck failed on p_field'); }
            if ( !(util.isString(p_op)) ) { throw new Error('typecheck failed on p_op'); }
        /* @DEVELOMPENT } */
        // cannot check compareTo, can be anything
        this._field = p_field;
        this._op = p_op;
        this._compareTo = p_compareTo;
    }
    this._className = 'sql.Condition';
}

Condition.prototype = new m_bjoo.BJOObject();
Condition.prototype.constructor = Condition;

Condition.prototype.field = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._field = field;
        return this;
    }
    return this._field;
};

Condition.prototype.op = function(op) {
    if ( typeof(op) !== 'undefined' ) {
        this._op = op;
        return this;
    }
    return this._op;
};

/**
 * @param compareTo field or value
 */
Condition.prototype.compareTo = function(compareTo) {
    if ( typeof(compareTo) !== 'undefined' ) {
        this._compareTo = compareTo;
        return this;
    }
    return this._compareTo;
};

Condition.prototype.compareToValue = function() {
    if ( this._compareTo instanceof m_sql_field.Field ) {
        return this._compareTo.value();
    } else {
        return this._compareTo;
    }
};

function condition(field, op, compareTo) { if ( arguments.length > 0 ) { return new Condition(field, op, compareTo); } else { return new Condition(); } }

function opFromName(name) { return _(Op).invert()[name]; }
function opNames() { return _(Op).keys(); }

function conditionsFromWeb(p_table, p_requestConditions) {
    if ( typeof(p_requestConditions) === 'undefined' || p_requestConditions === null ) {
        return [];
    }
    return _(p_requestConditions).map(function(c) {
        return condition(p_table.field(c.fieldName), opFromName(c.opName), c.value);
    });
}


exports.condition = condition;
exports.Op = Op;
exports.opFromName = opFromName;
exports.opNames = opNames;
exports.conditionsFromWeb = conditionsFromWeb;
