// sql/db.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of BJO2.
 *
 * BJO2 is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * factory for tthe db drivers
 */

var sqlite = require('sql/sqlite/db.js');

// url is currently actually just a filename, and sqlite is hardcoded... TODO
function DB(url) { 
    this._db = sqlite.db(url);
}

DB.prototype.open = function(url) {
    this._db.open(url);
    return this;
};

DB.prototype.runSql = function(queryString, params, callback) {
    return this._db.runSql(queryString, params, callback);
};

DB.prototype.allSql = function(queryString, params, callback) {
    return this._db.allSql(queryString, params, callback);
};

/* @param sql_query query
 */
DB.prototype.run = function(query, callback) {
    return this._db.run(query, callback);
}
DB.prototype.fetchRow = function(query, callback) {
    return this._db.fetchRow(query, callback);
};

DB.prototype.fetchAll = function(query, callback) {
    return this._db.fetchAll(query, callback);
};

DB.prototype.close = function() {
    return this._db.close();
};

function db(url) { return new DB(url); }

exports.DB = DB;
exports.db = db;
