// sql/db.js
/**
 * factory for tthe db drivers
 */

var sqlite = require('sql/sqlite/db.js');

// url is currently actually just a filename, and sqlite is hardcoded... TODO
function DB(url) { 
    this._db = sqlite.db(url);
}

DB.prototype.open = function(url) {
    this._db.open(url);
    return this;
};

DB.prototype.runSql = function(queryString, params, callback) {
    return this._db.runSql(queryString, params, callback);
};

DB.prototype.allSql = function(queryString, params, callback) {
    return this._db.allSql(queryString, params, callback);
};

/* @param sql_query query
 */
DB.prototype.run = function(query, callback) {
    return this._db.run(query, callback);
}
DB.prototype.fetchRow = function(query, callback) {
    return this._db.fetchRow(query, callback);
};

DB.prototype.fetchAll = function(query, callback) {
    return this._db.fetchAll(query, callback);
};

DB.prototype.close = function() {
    return this._db.close();
};

function db(url) { return new DB(url); }

exports.db = db;
