// sql/DDL.js
/*
 * Copyright (C) 2015-2017 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

const m_bjoo = require('BJOObject.js');

const Type = { create: 'create' };

function Query(type) {
    m_bjoo.BJOObject.apply(this);
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */ if ( arguments.length !== 1 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + 1); } /* @DEVELOMPENT } */
        this._type = type;
        this._table = null;
        this._field = null;
    }
}

Query.prototype = new m_bjoo.BJOObject();
Query.prototype.constructor = Query;

Query.prototype.type = function() {
    return this._type;
};
Query.prototype.table = function(table) {
    if ( typeof(table) !== 'undefined' ) {
        this._table = table;
        return this;
    }
    return this._table;
};

Query.prototype.field = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._field = field;
        return this;
    }
    return this._field;
};

function query(type) { return new Query(type); }
function create(table) { return new Query(Type.create).table(table); }

exports.Type = Type;
exports.Query = Query;
exports.query = query;
exports.create = create;
