// sql/ddl.js
var m_bjoo = require('BJOObject.js');

var Type = { create: 'create' };

function Query(type) {
    m_bjoo.BJOObject.apply(this);
    this._className = 'sql.ddl.Query';
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */ if ( arguments.length !== 1 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + 1); } /* @DEVELOMPENT } */
        this._type = type;
        this._table = null;
        this._field = null;
    }
}

Query.prototype = new m_bjoo.BJOObject();
Query.prototype.constructor = Query;

Query.prototype.type = function() {
    return this._type;
};
Query.prototype.table = function(table) {
    if ( typeof(table) !== 'undefined' ) {
        this._table = table;
        return this;
    }
    return this._table;
};

Query.prototype.field = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._field = field;
        return this;
    }
    return this._field;
};

function query(type) { return new Query(type); }
function create(table) { return new Query(Type.create).table(table); }

exports.Type = Type;
exports.Query = Query;
exports.query = query;
exports.create = create;
