// sql/field.js
"use strict";

var _ = require('underscore');
var util = require('util.js');
var bjoo = require('BJOObject.js');
var m_sql_fieldLink = require('sql/fieldLink.js');
var m_sql_field = require('sql/field.js');

var DataType = { int: 'int', string: 'string' }

function ValidationException(field, message) {
   this._field = field;
   this._message = message;
   this._className = "sql.ValidationException";
}

ValidationException.prototype = new bjoo.BJOObject();
ValidationException.prototype.constructor = ValidationException;

function Field(/*String*/name, /*m_sql_field.DataType*/dataType, /*[Object]*/value, /*[String]*/label) {
    bjoo.BJOObject.call(this);
    this._className = 'sql.Field';
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT */ if ( !(util.isString(name)) ) { throw new Error('typecheck failed'); }
        /* @DEVELOPMENT */ if ( !(util.isString(dataType)) ) { throw new Error('typecheck failed'); }
        this._name = name;
        this._dataType = dataType;
        this._table = null;
        if ( typeof(value) !== 'undefined' ) {
            this._value = value;
        } else {
            this._value = null;
        }
        if ( typeof(label) !== 'undefined' ) {
            this._label = label;
        } else {
            this._label = name;
        }
        this._links = {};
        this._seq = undefined; // seq nr, e.g. in a table
        this._isEditable = true; 
        this._isRequired = false; 
    }
}

Field.prototype = new bjoo.BJOObject();
Field.prototype.constructor = Field;

Field.prototype.clone = function() {
    var f = new Field(this._name, this._dataType, this._value);
    f.table(this._table);
    f.isEditable(this._isEditable);
    f.isRequired(this._isRequired);
    f.links(this._links);
    return f;
};

Field.prototype.table = function(table) {
	if ( typeof(table) !== 'undefined' ) {	
	    this._table = table;
        return this;
    }
    return this._table;
};

Field.prototype.seq = function(seq) {
	if ( typeof(seq) !== 'undefined' ) {	
	    this._seq = seq;
        return this;
    }
    return this._seq;
};

Field.prototype.name = function(name) {
	if ( typeof(name) !== 'undefined' ) {	
	    this._name = name;
        return this;
    }
    return this._name;
};

Field.prototype.isEditable = function(isEditable) {
	if ( typeof(isEditable) !== 'undefined' ) {	
	    this._isEditable = isEditable;
        return this;
    }
    return this._isEditable;
};

Field.prototype.isRequired = function(isRequired) {
	if ( typeof(isRequired) !== 'undefined' ) {	
	    this._isRequired = isRequired;
        return this;
    }
    return this._isRequired;
};

Field.prototype.identifierName = function() {
    return this._name; // TODO!!!
};

Field.prototype.fQName = function() {
    if ( this._table === null ) {
        return this.name();
    } else {
        return this._table.name() + '.' + this.name(); // TODO, at some point we'll need a unique table identifier
    }
};

Field.prototype.accessorName = function() {
    return this.identifierName();
};

Field.prototype.getterName = function() {
    var iName = this.identifierName();
    return 'get' + iName.substr(0, 1).toUpperCase() + iName.substr(1);
};

Field.prototype.setterName = function() {
    var iName = this.identifierName();
    return 'set' + iName.substr(0, 1).toUpperCase() + iName.substr(1);
};

Field.prototype.dataType = function(dataType) {
	if ( typeof(dataType) !== 'undefined' ) {	
	    this._dataType = dataType;
        return this;
    }
    return this._dataType;
};

Field.prototype.value = function(value) {
	if ( typeof(value) !== 'undefined' ) {	
	    this._value = value;
        return this;
    }
    return this._value;
};

Field.prototype.label = function(label) {
	if ( typeof(label) !== 'undefined' ) {	
	    this._label = label;
        return this;
    }
    return this._label;
};

/*
 * getter/setter 
 * field links for this field
 */
/*[]/this*/Field.prototype.links = function(links) {
    if ( typeof(links) !== 'undefined' ) {
        this._links = links;
        return this;
    }
    return this._links;
};

/*
 * a field is linked to another table via a fieldLink
 * - the table will be extracted from the link.field
 */
/*m_sql_fieldLink.FieldLink/this*/Field.prototype.link = function(/*m_sql_fieldLink.FieldLink / m_sql_table.Table*/linkOrTable) {
    if ( linkOrTable instanceof m_sql_fieldLink.FieldLink ) {
        this._links[linkOrTable.targetField().table().guid()] = linkOrTable;
        return this;
    }
    return this._links[linkOrTable.guid()];
};

/*private []*/Field.prototype._linksWebized = function() {
    return _(this._links).map(function(l) {
        return l.webize();
    });
};


// return non-circular object, so it can be jsonified and sent to the webz
// TODO name is a bit dodgy
// @pure
/*Object*/Field.prototype.webize = function() {
    return { 
        className: this._className, 
        name: this._name, 
        dataType: this._dataType, 
        value: this._value, 
        label: this._label,
        seq: this._seq,
        isEditable: this._isEditable,
        isRequired: this._isRequired,
        links: this._linksWebized()
    };
};

/* validate will not return a value, but rather throw a ValidationException on error */
Field.prototype.validate = /*void*/function(value, ctx) {
    // if no validation is set, we will assume all is good
    if ( typeof(this._validation) !== 'undefined' ) {
        this._validation.bind(this)(value, ctx); // ignore return value so user need not return this
    }
    return this;
};

Field.prototype.validation = function(/*function(value, [ctx])*/validation) {
	if ( typeof(validation) !== 'undefined' ) {	
	    this._validation = validation;
        return this;
    }
    return this._validation;
};


function field(name, dataType, value, label) { if ( arguments.length > 1 ) { return new Field(name, dataType, value, label); } else { return new Field(); } }
function validationException(field, message) { return new ValidationException(field, message); }

exports.Field = Field;
exports.field = field;
exports.DataType = DataType;
exports.ValidationException = ValidationException;
exports.validationException = ValidationException;
