// sql/Filter.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

const _ = require('underscore');
const m_bjoo = require('BJOObject.js');
const Util = require('Util.js');
const sql_Field = require('sql/Field.js');

const Op = { eq: 'eq', gt: 'gt', lt: 'lt', gte: 'gte', lte: 'lte', ne: 'ne', like: 'like' };

/**
 * @param {sql_Field.Field} field
 * @param {Op} op
 * @param {any} compareTo
 */
function Filter(field, op, compareTo) {
    m_bjoo.BJOObject.call(this);
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */ 
            if ( arguments.length !== 3 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + 3); }
            if ( !(field instanceof sql_Field.Field) ) { throw new Error('typecheck failed on field'); }
            if ( !(Util.isString(op)) ) { throw new Error('typecheck failed on op'); }
        /* @DEVELOMPENT } */
        // cannot check compareTo, can be anything
        this._field = field;
        this._op = op;
        this._compareTo = compareTo;
    }
}

function create(field, op, compareTo) { if ( arguments.length > 0 ) { return new Filter(field, op, compareTo); } else { return new Filter(); } }


Filter.prototype = new m_bjoo.BJOObject();
Filter.prototype.constructor = Filter;

Filter.prototype.getField = function() {
    return this._field;
};
Filter.prototype.setField = function(field) {
    this._field = field;
    return this;
};

Filter.prototype.getOp = function() {
    return this._op;
};
Filter.prototype.setOp = function(op) {
    this._op = op;
    return this;
};

Filter.prototype.getCompareTo = function() {
    return this._compareTo;
};
Filter.prototype.setCompareTo = function(/* field or value */ compareTo) {
    this._compareTo = compareTo;
    return this;
};

Filter.prototype.setCompareToValue = function() {
    if ( this._compareTo instanceof sql_Field.Field ) {
        return this._compareTo.value(); // TODO OOOOPS this will not fly!!!
    } else {
        return this._compareTo;
    }
};

function opFromName(name) { return _(Op).invert()[name]; }
function opNames() { return _(Op).keys(); }

function filtersFromWeb(table, requestFilters) {
    if ( typeof(requestFilters) === 'undefined' || requestFilters === null ) {
        return [];
    }
    return _(requestFilters).map(function(c) {
        return filter(table.field(c.fieldName), opFromName(c.opName), c.value);
    });
}


exports.create = create;
exports.Op = Op;
exports.opFromName = opFromName;
exports.opNames = opNames;
exports.filtersFromWeb = filtersFromWeb;
