// sql/LookupField.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * Mapitor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * this is a lookup field onto another entity's id
 * note that we cannot lookup onto any other field
 * we shall probably create a better version of this later on...
 */

"use strict";

const _ = require('lodash');
const q = require('q');

const util = require('util');
const Util = require('Util.js');
const model_EntitySet = require('entity/EntitySet.js');
const sql_Field = require('sql/Field.js');
const sql_ValueField = require('sql/ValueField.js');
const sql_Query = require('sql/Query.js');

function LookupField() {
    sql_ValueField.ValueField.call(this);
}

/**
 * @param name the name of this field
 * @param idField the field by which we link to the "other" entity
 * @param EntitySet the "other" entity set
 */
function create(/* String */ name, /* sql_LookupIdField.LookupIdField */ idField, /* entity_EntitySet.EntitySet */ entitySet, /* String */ label) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            if ( arguments.length !== 4 ) { throw new Error('arguments.length must be 4'); }
        /* @DEVELOMPENT } */
        const field = new LookupField();
        field.setName(name);
        // ??? TODO ??? 
        field.setDataType(sql_Field.DataType.int);
        field.setIdField(idField);
        field.setEntitySet(entitySet);
        field.setLabel(label);
        field.setIsPersistent(false);
        return field;
    }  else {
        return new LookupField();
    }
}

exports.LookupField = LookupField;
exports.create = create;

LookupField.prototype = new sql_ValueField.ValueField();
LookupField.prototype.constructor = LookupField;

LookupField.prototype.clone = function() {
    return create(this._name, this._idField, this._entitySet, this._label);
};

/* getter/setter */

LookupField.prototype.getIdField = function() {
    return this._idField;
};
LookupField.prototype.setIdField = function(/* sql_LookupIdField */ idField) {
    this._idField = idField;
    return this;
};

LookupField.prototype.getEntitySet = function() {
    return this._entitySet;
};

LookupField.prototype.setEntitySet = function(/* entity_EntitySet.EntitySet */ entitySet) {
    this._entitySet = entitySet;
    return this;
};

LookupField.prototype.getValue = function() {
    return this._idField.getValue().then( (id) => {
        if ( !id ) { throw new Error('no id for lookup field ' + this._name + ' (' + this._idField.getName() + ')'); }
        return this._entitySet.loadEntityById(id);
    });
};

LookupField.prototype.setValue = function(/* entity_Entity */ entity) {
    sql_ValueField.ValueField.call(this, entity);
    this._idField.setValue(entity.getId());
};


/* others */

LookupField.prototype.updateReferences = function() {
    this.setIdField(this._table.getField(this._idField.getName()));
};
