// sql/Table.js
"use strict";

var bjoo = require('BJOObject.js');
var _ = require('underscore');
var m_sql_field = require('sql/field.js');

function Table(name) {
    this._className = 'sql.Table';
    if ( arguments.length > 0 ) {
        this._name = name;
        this._fields = {};
        this._indices = {};
    }
}

Table.prototype = new bjoo.BJOObject();
Table.prototype.constructor = Table;

Table.prototype.name = function(name) {
    if ( typeof(name) !== 'undefined' ) {
        this._name = name;
        return this;
    }
    return this._name;
};

Table.prototype.field = function(nameOrField) {
    if ( nameOrField instanceof m_sql_field.Field ) { 
        nameOrField.table(this);
        nameOrField.seq(_(this._fields).values().length);
        this._fields[nameOrField.name()] = nameOrField;
        return this;
    }
    return this._fields[nameOrField];
};

Table.prototype.fields = function(fields) {
    if ( typeof(fields) !== 'undefined' ) {
        _.each(fields, function(f) { this.field(f); }, this);
	return this;
    }
    return this._fields;
};

Table.prototype.fieldsAsList = function() {
    return _(this._fields).values();
};


Table.prototype.index = function(nameOrIndex) {
    if ( typeof(nameOrField.className) !== 'undefined' && (nameOrField.className === 'sql.CalcField' || nameOrField.className() === 'sql.LookupField' || nameOrField.className() === 'sql.Field') ) {
        nameOrField.table(this);
        this._indices[nameOrField.name()] = nameOrIndex;
        return this;
    }
    return this._indices[nameOrField];
};

Table.prototype.indices = function(indices) {
    if ( typeof(indices) !== 'undefined' ) {
        _.each(indices, function(f) { this.index(f); }, this);
	return this;
    }
    return this._indices;
};

Table.prototype.fieldLinks = function(p_fieldLinks) {
    if ( typeof(p_fieldLinks) !== 'undefined' ) {
        this._fieldLinks = p_fieldLinks;
        return this;
    }
    return this._fieldLinks;
};

function table(name) { return new Table(name); }

exports.Table = Table;
exports.table = table;
