// sql/Table.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

var bjoo = require('BJOObject.js');
var _ = require('underscore');
var sql_Field = require('sql/Field.js');
var Errors = require('Errors.js');

function Table(name) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */ if ( arguments.length !== 1 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be 1'); } /* @DEVELOMPENT } */ 
        this._name = name; 
        this._fields = {};
        this._indices = {};
        this._labelFields = []; // we use an array here because we will never need to access them via name
    }
}
function create(name) { return new Table(name); }

Table.prototype = new bjoo.BJOObject();
Table.prototype.constructor = Table;

Table.prototype.getName = function() {
    return this._name;
};
Table.prototype.setName = function(name) {
    this._name = name;
    return this;
};

Table.prototype.getField = function(name) {
    return this._fields[name];
};

Table.prototype.addField = function(field) {
    /* @DEVELOPMENT */ if ( !(field instanceof sql_Field.Field) ) { throw new Errors.TypeError('field must be a sql_Field.Field'); }
    field.setTable(this);
    field.setSeq(_(this._fields).values().length);
    this._fields[field.getName()] = field;
    return this;
};

Table.prototype.getFields = function(fields) {
    return this._fields;
};
Table.prototype.setFields = function(fields) {
    /* @DEVELOPMENT */ if (!(fields instanceof Array)) { throw new Error('fields must be an array'); }
    _.each(fields, function(f) { this.field(f); }, this);
    return this;
};

Table.prototype.getFieldsAsList = function() {
    return _(this._fields).values();
};

/**
 * fields to present for labelling the result in a lookup field
 */
Table.prototype.getLabelFields = function() {
    return this._labelFields;
};
Table.prototype.setLabelFields = function(/* array(sql_Field.Field) */ labelFields) {
    this._labelFields = labelFields;
    return this;
};


Table.prototype.index = function(nameOrIndex) {
    if ( nameOrInstance instanceof sql_Field.Field ) {
        nameOrField.table(this);
        this._indices[nameOrField.name()] = nameOrIndex;
        return this;
    }
    return this._indices[nameOrField];
};

Table.prototype.indices = function(indices) {
    if ( typeof(indices) !== 'undefined' ) {
        _.each(indices, function(f) { this.index(f); }, this);
	return this;
    }
    return this._indices;
};

Table.prototype.getFieldLinks = function() {
    return this._fieldLinks;
};
Table.prototype.setFfieldLinks = function(fieldLinks) {
    this._fieldLinks = fieldLinks;
    return this;
};

/*
 * { lookup: { module, controller, action, labelFieldName } }
 */
Table.prototype.frontendSettings = function(/* optional Object */ frontendSettings) {
    if ( typeof(frontendSettings) !== 'undefined' ) {
        this._frontendSettings = frontendSettings;
        return this;
    }
    return this._frontendSettings;
};



exports.Table = Table;
exports.create = create;
