// entity/Entity.js
/*
 * Copyright (C) 2015-2017 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

/**
 * an Entity is an object that is fit for its corresponding entry in the database
 * ... apart from that, nothing is known about the entity, so for saving it, we shall 
 * have to use all primary key fields
 *
 * TODO: filename is lowercase now, and damn windoze!!
 */

"use strict";

const _ = require('lodash');
const q = require('q');
const bjoo = require('BJOObject.js');
const model_EntityModel = require('model/EntityModel.js');

/**
 * The generic Entity class
 * @class entity.Entity
 * @constructor
 */
function Entity() {
    bjoo.BJOObject.call(this);
}

/**
 * external constructor
 * @function create
 * @static
 * @param {model.EntityModel} model
 */
function create(/* model_EntityModel.Entity */ model) {
    const entity = new Entity();
    if ( arguments.length > 0 ) {
        /* @DEVELOMPENT { */
            if ( !(model instanceof model_EntityModel.EntityModel) ) { throw new Error('model must be of type model_EntityModel.EntityModel:' + model); }
        /* @DEVELOMPENT } */
        entity.setModel(model);
    }
    return entity;
}

Entity.prototype = bjoo.create();
Entity.prototype.constructor = Entity;

Entity.prototype.setModel = function(/* model_EntityModel.Entity */ model) {
    /* @DEVELOMPENT { */
        /* TODO: for testing, we could create an "interface" an then derive a mock object from that */
        // if ( !(model instanceof model_EntityModel.IEntityModel) ) { throw new Error('model must be of type model_EntityModel.IEntityModel'); }
        if ( this._model ) {
            console.warn('this._model is already set in Entity.setModel, some attributes might remain');
        }
    /* @DEVELOMPENT } */
    this._model = model;
    _.each(model.getAttributeNames(), ( name ) => {
        this._createAccessors(name);
    });
};

/**
 * create getters and setters for an attribute
 *
 * @method _createAccessors
 * @private
 * @param {String} key the name of the attribute
 */
Entity.prototype._createAccessors = function(key) {
    this['get' + key.substr(0, 1).toUpperCase() +  key.substr(1)] = function() { return this._model.getAttributeValue(key); }.bind(this);
    this['set' + key.substr(0, 1).toUpperCase() +  key.substr(1)] = function(val) { this._model.setAttributeValue(key, val); return this; }.bind(this);
};

/**
 * save the entity to the db
 * @method save
 * @async
 * @return {Promise}
 */
Entity.prototype.save = function() {
    return this._model.save();
};

/**
 * delete the entity from the db
 * @method delete
 * @async
 * @return {Promise}
 */
Entity.prototype.delete = function() {
    return this._model.delete();
};

/**
 * get all fields as a list (array)
 *
 * @method getFieldsAsList
 * @return {Array}
 */
Entity.prototype.getFieldsAsList = function() {
    return this._model.getFieldsAsList();
};

/**
 * @method getValuesAsObject
 * @param {Array(String)} [names] get only the values for names in this list
 * @async
 * @return {Promise(Object)}
 */
Entity.prototype.getValuesAsObject = Entity.prototype.get = function(names) {
    if ( names === undefined ) {
        names = this._model.getAttributeNames();
    }

    let promisedObj = {};
    return q.all(_.map(names, ( name ) => {
        return this._model.getAttributeValue(name).then( ( value ) => {
            promisedObj[name] = value;
        });
    })).then( () => {
        return promisedObj;
    });
};

/**
 *  get field for name
 *  @method getField
 *  @param {String} name
 *  @return {sql.Field}
 */
Entity.prototype.getField = function(name) {
    return this._model.getAttribute(name);
};

/**
 * validate the entity, throwing an exception to the promise if invalid
 * @method validate
 * @async
 * @return {Promise}
 */
Entity.prototype.validate = function() {
    return this._model.validate();
};

/**
 * set a validation function for this entity
 *
 * @method setValidation
 * @param {Function(context)} validation
 */
Entity.prototype.setValidation = function(/* function(ctx) */ validation) {
    this._validation = validation;
};

/** 
 * return an object that is not cyclic and contains all necessary fields for a json representation
 *
 * @method toJson
 * @return {Promise(Object)}
 */ 
Entity.prototype.toJson = function() {
    return q.all(_.map(this.getFieldsAsList(), ( field ) => {
        return field.toJson();
    })).then( (fields) => {
        return { fields: fields };
    });
}

exports.Entity = Entity;
exports.create = create;
