// server/dao/fieldLink.js
var Type = { oneToMany: '1:n', oneToOne: '1:1', manyToMany: 'n:m' };

function FieldLink(sourceField, targetField, type) {
    this._sourceField = sourceField;
    this._targetField = targetField;
    if ( typeof(type) === 'undefined' ) {
        type = Type.oneToMany;
    }
    this._type = type;
}

FieldLink.prototype.sourceField = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._sourceField = field;
        return this;
    }
    return this._sourceField;
};

FieldLink.prototype.targetField = function(field) {
    if ( typeof(field) !== 'undefined' ) {
        this._targetField = field;
        return this;
    }
    return this._targetField;
};

FieldLink.prototype.type = function(type) {
    if ( typeof(type) !== 'undefined' ) {
        this._type = type;
        return this;
    }
    return this._type;
};

FieldLink.prototype.createSelect = function(aggregate) {
    return query.select()
        .aggregate(aggregate)
        .from(this.targetField().table())
        .where(condition.condition(this.targetField(), condition.Op.eq, this.sourceField()));
};

function fieldLink(sourceField, targetField, type) { return new FieldLink(sourceField, targetField, type); }

exports.FieldLink = FieldLink;
exports.fieldLink = fieldLink;
exports.Type = Type;
