// sql/sqlite/db.js
var sqlite3 = require('sqlite3').verbose();
var query = require('./query.js');

function DB() {
    this._db = null;
}

DB.prototype.open = function(fileName) {
    this._db = new sqlite3.Database(fileName);
    return this;
};

DB.prototype.close = function() {
    this._db.close();
    return this;
};

DB.prototype.runSql = function(sqlString, params, callback) {
    console.log('sqlite_db_runSql', sqlString, params);
    this._db.run(sqlString, params, function(err) {
        callback(err);
    });
    return this;
};

DB.prototype.run = function(sqlQuery, params, callback) {
    var q = query.query(sqlQuery);
    var qs = q.queryString();
    var p = q.params();
    console.log('sqlite_db_run: queryString: ', qs, p);
    this.runSql(qs, p, function(err) {
        callback(err);
    });
    return this;
};


/**
 * string query, array params -> callback(err, row)
 */
DB.prototype.fetchRow = function(sqlQuery, callback) {
    var q = query.query(sqlQuery);
    var qs = q.queryString();
    var p = q.params();
    console.log('sqlite_db_fetchRow: queryString is', qs, p);
    this._db.get(qs, p, function(err, row) {
        console.log('sqlite_db_fetchRow: result row is', row);
        callback(err, row);
    });
    return this;
};

function db(fileName) {
    return new DB(fileName);
}

exports.db = db;
