// app/customer/customerBo.js
var util = require('util');
var m_bo = require('bo/bo.js');
var m_app_customer_customerDao = require('app/customer/CustomerDao.js');

function CustomerBo(p_db) {
    m_bo.BO.call(this, p_db, m_app_customer_customerDao.customerDao(p_db));
    this._className = 'customerBo.CustomerBo';
};

CustomerBo.prototype = new m_bo.BO();

////////////////////////////////////////////////////

CustomerBo.prototype.id = function(id) {
    if ( typeof(id) !== 'undefined' ) {
        this._fields['id'].value(id);
        return this;
    }
    return this._fields['id'].value();
};

CustomerBo.prototype.firstName = function(firstName) {
    if ( typeof(firstName) !== 'undefined' ) {
        this._fields['firstName'].value(firstName)
        return this;
    }
    return this._fields['firstName'].value();
};

CustomerBo.prototype.lastName = function(lastName) {
    if ( typeof(lastName) !== 'undefined' ) {
        this._fields['lastName'].value(lastName);
        return this;
    }
    return this._fields['lastName'].value();
};

CustomerBo.prototype.loadById = function(id, callback) {
    this._dao.loadById(id, function(err, dao) {
        if (err) callback(err);
        this._fieldValuesFromDao(dao);
        callback(false, this);
    }.bind(this));
};

function customerBo(p_db) { return new CustomerBo(p_db); }

exports.customerBo = customerBo;
exports.CustomerBo = CustomerBo;
