// dao/dao.js
"use strict";

var table = require('sql/table.js');
var query = require('sql/query.js');
var field = require('sql/field.js');
var condition = require('sql/condition.js');
var sqliteQuery = require('sql/sqlite/query.js');
var _ = require('underscore');

var DAO = function(p_db, p_table) {
    if ( arguments.length > 0 ) {
        this._className = 'dao.DAO';
        this._db = p_db;
        this._fields = {};
        if ( typeof(p_table) === 'undefined' ) {
            throw new Error("can't construct dao without a table");
        }
        this.table(p_table);
        this._addAccessors();
    }
};

DAO.prototype._addAccessors = function() {
    // TODO: name must be identifier-compatible!
    _.each(this._fields, function(f) { 
        this[f.accessorName()] = function(val) { 
            if ( arguments.length === 1 && typeof(val) !== 'undefined' ) {
                return this.fieldValue(f.name(), val); 
            } else if ( arguments.length === 0 ) {
                return this.fieldValue(f.name()); 
            }
        }.bind(this); 
    }, this);
    return this;
};

DAO.prototype.db = function(db) {
    this._db = db;
    return this;
};

DAO.prototype.table = function(table) {
    if ( typeof(table) !== 'undefined' ) {
        this._table = table;
        this._fieldsFromTable(table);
        return this;
    }
    return this._table;
};

DAO.prototype.field = function(fieldOrName) {
    if ( typeof(fieldOrName.className) !== 'undefined' && fieldOrName.className() === 'sql.Field' ) {
        fieldOrName.table(this.table());
        this._fields[fieldOrName.name()] = fieldOrName;
        return this;
    }
    return this._fields[fieldOrName];
};

DAO.prototype.addFields = function(fields) {
    var self = this;
    _.each(fields, function(f) { self.field(f); });
};

DAO.prototype.fieldValue = function(name, value) {
    if ( typeof(value) !== 'undefined' ) {
        this.field(name).value(value);
        return this;
    }
    return this._fields[name].value();
};

DAO.prototype.fieldsAsList = function() {
    return _.values(this._fields);
};

DAO.prototype._fieldsFromTable = function(table) {
    var tf = table.fields();
    _.each(_.keys(tf), function(key) { this.field(tf[key]); }, this);
};

DAO.prototype._setValuesFromRow = function(row) {
    _.each(this._fields, function(f) { 
        if ( typeof(row[f.name()]) !== 'undefined' ) {
            f.value(row[f.name()]);
        }
    }, this);
    return this;
}

DAO.prototype.loadOneByQuery = function(query, callback) {
    this._db.fetchRow(query, function(err, row) {
        if ( err ) return callback(err);
        this._setValuesFromRow(row);
        callback(false, this);
    }.bind(this));
};


function dao(db, table) { return new DAO(db, table); }

exports.DAO = DAO;
exports.dao = dao;
