// dao/dao.js
"use strict";

var async = require('async');
var _ = require('underscore');
var table = require('sql/table.js');
var query = require('sql/query.js');
var field = require('sql/field.js');
var m_sql_condition = require('sql/condition.js');
var m_sql_calcField = require('sql/calcField.js');
var condition = require('sql/condition.js');
var m_sql_query = require('sql/query.js');
var m_sql_aggregate = require('sql/aggregate.js');
var m_dao_daoDataProvider = require('dao/daoDataProvider.js');
var m_bjoo = require('BJOObject.js');

var DAO = function(/*Db*/p_db, /*Table*/p_table) {
    m_dao_daoDataProvider.DaoDataProvider.call(this, p_db, p_table);
    if ( arguments.length > 0 ) {
        /* #ifdef DEVELOPMENT */ if ( arguments.length !== 2 ) { console.log('arguments.length', arguments.length); throw new Error('Illegal Arguments Length'); }
        this._className = 'dao.DAO';
        this._db = p_db;
        this._fieldLinks = [];
        this._addAccessors();
    }
};

DAO.prototype = new m_dao_daoDataProvider.DaoDataProvider();
DAO.prototype.constructor = DAO;

DAO.prototype._addAccessors = function() {
    // TODO: name must be identifier-compatible!
    _.each(this._fields, function(f) { 
        this[f.accessorName()] = function(val) { 
            if ( arguments.length === 1 && typeof(val) !== 'undefined' ) {
                return this.fieldValue(f.name(), val); 
            } else if ( arguments.length === 0 ) {
                return this.fieldValue(f.name()); 
            }
        }.bind(this); 
    }, this);
    return this;
};

DAO.prototype.db = function(db) {
    this._db = db;
    return this;
};

DAO.prototype.fieldLinks = function(p_fieldLinks) {
    if ( typeof(p_fieldLinks) !== 'undefined' ) {
        this._fieldLinks = p_fieldLinks;
        return this;
    }
    return this._fieldLinks;
};

DAO.prototype.fieldValue = function(name, value) {
    if ( typeof(value) !== 'undefined' ) {
        this.field(name).value(value);
        return this;
    }
    return this._fields[name].value();
};

DAO.prototype._setValuesFromRow = function(row) {
    _.each(this._fields, function(f) { 
        if ( typeof(row[f.name()]) !== 'undefined' ) {
            f.value(row[f.name()]);
        }
    }, this);
    return this;
};

DAO.prototype.fieldValuesFromBo = function(bo) {
    _(this.writableFieldsAsList()).each( function(f) { 
        this.fieldValue(f.name(), bo.fieldValue(f.name()));
    }, this);
    return this;
};

/* low level function, will not perform calculateCalcFields()! */
DAO.prototype.loadOneByQuery = function(query, callback) {
    this._db.fetchRow(query, function(err, row) {
        if ( err ) return callback(err);
        this._setValuesFromRow(row);
        callback(false, this);
    }.bind(this));
};

DAO.prototype.calculateCalcFields = function(conditions, callback) {
    _(this.calcFieldsAsList()).each(function(f) { f.value(null); });
    async.each(this.calcFieldsAsList(), 
        function(f, callback2) {
            f.conditions(conditions);
            var q = f.query();
            this._db.fetchRow(q, function(err, row) {
                if ( err ) return callback(err);
                f.value(row[f.name()]);
                callback2(false, this);
            }.bind(this));
        }.bind(this),
        function(err) {
            callback(err, this);
        }.bind(this)
    );
};

DAO.prototype.countByConditions = function(conditions, callback) {
    var q = m_sql_query.select()
        .fields([m_sql_aggregate.aggregate(m_sql_aggregate.Type.count)])
        .from(this._table);
    if ( conditions.length > 0 ) {
        q.where(conditions);
    }
    this._db.fetchRow(q, function(err, row) {
        if ( err ) return callback(err);
        callback(false, row.aggregate);
    });

};

function dao(db, table) { return new DAO(db, table); }

exports.DAO = DAO;
exports.dao = dao;
