// dao/dao.js
var table = require('sql/table.js');
var query = require('sql/query.js');
var field = require('sql/field.js');
var condition = require('sql/condition.js');
var sqliteQuery = require('sql/sqlite/query.js');
var _ = require('underscore');

var DAO = function(db, table) {
    this._className = 'dao.DAO';
    this._db = db;
    this._fields = {};
    this._table = null;
    this.table(table);
    this._addGetters();
};

DAO.prototype._addGetters = function() {
    // TODO: name must be identifier-compatible!
    _.each(this._fields, function(f) { this[f.getterName()] = function() { console.log('inner', this); return this.fieldValue(f.name()); }.bind(this); }.bind(this));
};

DAO.prototype.db = function(db) {
    this._db = db;
    return this;
};

DAO.prototype.table = function(table) {
    if ( typeof(table) !== 'undefined' ) {
        this._table = table;
        this._fieldsFromTable(table);
        return this;
    }
    return this._table;
};

DAO.prototype.field = function(fieldOrName) {
    if ( typeof(fieldOrName.className) !== 'undefined' && fieldOrName.className() === 'sql.Field' ) {
        console.log('dao is adding field ', fieldOrName.name(), fieldOrName.value());
        fieldOrName.table(this.table());
        this._fields[fieldOrName.name()] = fieldOrName;
        console.log('dao has added field ', fieldOrName.name(), fieldOrName.value());
        return this;
    }
    return this._fields[fieldOrName];
};

DAO.prototype.addFields = function(fields) {
    var self = this;
    _.each(fields, function(f) { self.field(f); });
};

DAO.prototype.fieldValue = function(name, value) {
    if ( typeof(value) !== 'undefined' ) {
        this.field(name).value(value);
        return this;
    }
    return this._fields[name].value();
};

DAO.prototype._fieldsAsList = function() {
    return _.values(this._fields);
};

DAO.prototype._fieldsFromTable = function(table) {
    console.log('f d d', _.values(table.fields()));
    var self = this;
    // _.map(_.values(table.fields()), function(f) { console.log('f d d is adding field', f.name(), f.value(), f.name, f.value); self.field(f); }, this);
    var key;
    var f = table.fields();
    console.log('START FOREACH');
    for (key in f) {
        var ff = f[key];
        console.log(ff.name(), ff.value(), ff._name, ff._value);
        this.field(ff);
    }
    console.log('END FOREACH');
};


function dao(db, table) { return new DAO(db, table); }

exports.DAO = DAO;
exports.dao = dao;
