// dao/daoDataProvider.js
/*
 * Copyright (C) 2015 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of BJO2.
 *
 * Mapitor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

var async = require('async');
var _ = require('underscore');
var m_sql_table = require('sql/table.js');
var query = require('sql/query.js');
var m_sql_field = require('sql/field.js');
var m_sql_lookupField = require('sql/lookupField.js');
var m_sql_db = require('sql/db.js');

/*
 * abstract provider for dao data - base class for DaoSet and DAo
 */

var m_bjoo = require('BJOObject.js');

function DaoDataProvider() {
    m_bjoo.BJOObject.call(this);
    this._className = 'dao.DaoDataProvider';
    /* @DEVELOPMENT { */ if ( arguments.length > 0 ) { throw new Error('Illegal argument length '  + arguments.length + ', should be 0'); }  /* @DEVELOMPENT } */ 
    this._fields = {};
};

DaoDataProvider.prototype = new m_bjoo.BJOObject();
DaoDataProvider.prototype.constructor = DaoDataProvider;


DaoDataProvider.prototype.table = function(p_table) {
    if ( typeof(p_table) !== 'undefined' ) {
        /* @DEVELOPMENT { */ if ( !(p_table instanceof m_sql_table.Table) ) { throw new Error('typecheck failed on p_table: ' + p_table); }  /* @DEVELOMPENT } */
        this._table = p_table;
        this._fieldsFromTable(p_table);
        return this;
    }
    return this._table;
};

DaoDataProvider.prototype.field = function(fieldOrName) {
    if ( fieldOrName instanceof m_sql_field.Field ) { // TODO we really need to genericize this!!!!!!!!
        fieldOrName.table(this.table());
        this._fields[fieldOrName.name()] = fieldOrName;
        return this;
    }
    return this._fields[fieldOrName];
};

DaoDataProvider.prototype.addFields = function(fields) {
    var self = this;
    _.each(fields, function(f) { self.field(f); });
};

DaoDataProvider.prototype._fieldsFromTable = function(table) {
    var tf = table.fields();
    _.each(_.keys(tf), function(key) { 
        this.field(tf[key].clone()); 
    }, this);
};


DaoDataProvider.prototype.fieldsAsList = function() {
    return _.values(this._fields);
};

DaoDataProvider.prototype.writableFieldsAsList = function() {
    return _(this._fields).values().filter(function(f) { return f.className() !== 'sql.CalcField'; }); // TODO this condition is not exactly right
};

DaoDataProvider.prototype.databaseFieldsAsList = function() {
    return _(this._fields).values().filter(function(f) { return f.className() !== 'sql.CalcField'; }); // TODO this condition is not exactly right
};

DaoDataProvider.prototype.calcFieldsAsList = function() {
    return _(this._fields).values().filter(function(f) { return (f.className() === 'sql.CalcField'); }); 
}

DaoDataProvider.prototype.lookupFieldsAsList = function() {
    return _(this._fields).values().filter(function(f) { return (f.className() === 'sql.LookupField'); }); 
}


function daoDataProvider(p_db, p_table) { if ( arguments.length > 0 ) { return new DaoDataProvider(p_db, p_table); } else { return new DaoDataProvider(); } }

exports.DaoDataProvider = DaoDataProvider;
exports.daoDataProvider = daoDataProvider;
