// dao/daoSet.js
/*
 * NOTE: if you use an "abstract dao", you need to explicitly set the table to the daoSet!
 * because otherwise, we cannot create a query for conditions, because we need the fields for that
 */

"use strict";

var _ = require('underscore');
var async = require('async');
var m_bjoo = require('BJOObject.js');
var m_dao_daoDataProvider = require('dao/daoDataProvider.js');
var m_dao_primaryDao = require('dao/primaryDao.js');
var m_dao_dao = require('dao/dao.js');
var m_sql_query = require('sql/query.js');
var m_sql_aggregate = require('sql/aggregate.js');
var table = require('sql/table.js');
var query = require('sql/query.js');
var field = require('sql/field.js');
var m_sql_condition = require('sql/condition.js');
var m_sql_table = require('sql/table.js');
var m_sql_db = require('sql/db.js');

function DaoSet(/* m_sql_db.DB */ p_db, /*Function*/ p_constructor) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT */ 
            var ARGLEN = 2; if ( arguments.length !== ARGLEN ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + ARGLEN); } 
            if ( !(p_db instanceof m_sql_db.DB) ) { throw new Error('typecheck failed on p_db: ' + p_db); }
            if ( !(p_constructor instanceof Function) ) { throw new Error('typecheck failed on p_constructor: ' + p_constructor); }
        /* @DEVELOMPENT } */
        m_dao_daoDataProvider.DaoDataProvider.call(this);
        this._db = p_db;
        this._table = undefined;
        this._className = 'dao.DaoSet';
        this._constructor = p_constructor;
    } else {
        m_dao_daoDataProvider.DaoDataProvider.call(this);
    }
};

function daoSet(/* m_sql_db.DB */ p_db, /*Function*/ p_constructor) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT */ 
            var ARGLEN = 2; if ( arguments.length !== ARGLEN ) { throw new Error('Illegal argument length '  + arguments.length + ', should be ' + ARGLEN); } 
            if ( !(p_db instanceof m_sql_db.DB) ) { throw new Error('typecheck failed on p_db: ' + p_db); }
            if ( !(p_constructor instanceof Function) ) { throw new Error('typecheck failed on p_constructor: ' + p_constructor); }
        /* @DEVELOMPENT } */
        return new DaoSet(p_db, p_constructor);
    } else {
        return new DaoSet();
    }
};

DaoSet.prototype = new m_dao_daoDataProvider.DaoDataProvider();
DaoSet.prototype.constructor = DaoSet;


/**
 * returns a pure, uninitialized dao
 */
/*dao.DAO*/ DaoSet.prototype.templateDao = function() {
    return this._constructor.call({}, this._db);
};

/**
 * if we have a table, we want to set it to the dao
 *
 * NOTE: this is only useful for "abstract" daos; for "application" daos that are usually primary daos, it would actually be destructive because
 * it might destroy the id field, probably
 */

DaoSet.prototype.tableToDao = function(/* m_dao_dao.DAO */ p_dao) {
    /* @DEVELOPMENT { */ var PARAM = this._table; if ( !(PARAM instanceof m_sql_table.Table || PARAM === undefined) ) { throw new Error('typecheck failed on ' + PARAM + ': ' + PARAM); }  /* @DEVELOMPENT } */
    /* @DEVELOPMENT { */ var PARAM = p_dao; if ( !(PARAM instanceof m_dao_dao.DAO) ) { throw new Error('typecheck failed on ' + PARAM + ': ' + PARAM); }  /* @DEVELOMPENT } */

    if ( this._table !== undefined ) {
        p_dao.table(this._table);
    }
};

/* low level function, will not perform calculateCalcFields()! */
DaoSet.prototype.loadOneByQuery = function(query, callback) {
    this._db.fetchRow(query, function(err, row) {
        if ( err ) return callback(err);
        var dao = this._constructor.call({}, this._db);
        this.tableToDao(dao);
        dao._setValuesFromRow(row);
        callback(false, dao);
    }.bind(this));
};

/* low level function, will not perform calculateCalcFields()! */
DaoSet.prototype.loadAllByQuery = function(query, callback) {
    this._db.fetchAll(query, function(err, rows) {
        /* @DEVELOPMENT { */ if ( err ) throw new Error(err); /* @DEVELOMPENT } */ if ( err ) return callback(err);
        var daos = [];
        _(rows).each(function(row) {
            var dao = this._constructor.call({}, this._db);
            this.tableToDao(dao);
            dao._setValuesFromRow(row);
            daos.push(dao);
        }.bind(this));
        callback(false, daos);
    }.bind(this));
};

/**
 * note: this assumes that the dao is a primary dao for calculating calc fields!
 */
DaoSet.prototype.loadAllByConditions = function(conditions, callback, calculateDetailCalcFields, calculateAggregatedCalcFields) {
    var self = this;
    /*
     * if we were created for an "abstract dao", whose table was explicitly set from the outside, then 
     * we cannot fetch the table from the template dao because its constructor has no knowledge of said table
     * so in this case, the user has to set our table instead
     */
    if ( this._table === undefined ) {
        var templateDao = this.templateDao();
        this.table(templateDao.table());
    } 
    var fields = this.databaseFieldsAsList();
    var q = m_sql_query.select()
                .fields(fields)
                .from(this._table);
    if ( conditions.length > 0 ) {
        q.conditions(conditions);
    }

    function loadDetailRows(err, aggregateDao) {
        if ( err ) { throw new Error('loadDetailRows failed'); }

        calculateDetailCalcFields &= ( self._constructor.call({}) instanceof m_dao_primaryDao.PrimaryDao ); // determine whether an object instantiated by us would be primary or not

        return self.loadAllByQuery(q, function(err, daos) {
            if (err) { console.error('error in loadAllByQuery-loadDetailRows', err); return callback(err); }
            if ( calculateDetailCalcFields ) {
                async.each(daos, 
                    function(dao, callback2) {
                        var singleBoConditions = conditions.slice();
                        singleBoConditions.push(m_sql_condition.condition(dao.field('id'), m_sql_condition.Op.eq, dao.field('id').value()));
                        async.parallel([
                            function(callback3) { dao.calculateCalcFields(singleBoConditions, callback3); },
                            function(callback3) { 
                                dao.populateLookupFields(function(err, dao) {
                                    callback3(err, dao); 
                                });
                            }
                        ], function(err) { 
                            callback2(err, daos, aggregateDao); 
                        });
                    }, function(err) {
                        callback(false, daos, aggregateDao);
                    });
            } else {
                async.each(daos, 
                    function(dao, callback2) {
                        dao.populateLookupFields(callback2);
                    }, function(err) {
                        callback(false, daos, aggregateDao);
                    }
                );
            }
        });
    }

    var aggregateDao = self.templateDao();
    if ( calculateAggregatedCalcFields ) {
        aggregateDao.calculateCalcFields(conditions, loadDetailRows);
    } else {
        loadDetailRows(false, undefined);
    }
};

DaoSet.prototype.calculateCalcFields = function(conditions, callback) {
    _(this.calcFieldsAsList()).each(function(f) { f.value(null); });
    async.each(this.calcFieldsAsList(), 
        function(f, callback2) {
            f.conditions(conditions);
            var q = f.query();
            this._db.fetchRow(q, function(err, row) {
                if ( err ) return callback(err);
                f.value(row[f.name()]);
                callback2(false, this);
            }.bind(this));
        }.bind(this),
        function(err) {
            callback(err, this);
        }.bind(this)
    );
};

DaoSet.prototype.countByConditions = function(conditions, callback) {
    var q = m_sql_query.select()
        .fields([m_sql_aggregate.aggregate(m_sql_aggregate.Type.count)])
        .from(this._table);
    if ( conditions.length > 0 ) {
        q.conditions(conditions);
    }
    this._db.fetchRow(q, function(err, row) {
        if ( err ) return callback(err);
        callback(false, row.aggregate);
    });

};

exports.DaoSet = DaoSet;
exports.daoSet = daoSet;
