// server/app/invoice/invoice.js
/*
 * Copyright (C) 2015,2016 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of SteerGlance.
 *
 * SteerGlance is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

const m_invoiceBo = require('app/invoice/invoiceBo.js');
const m_dao_daoSet = require('dao/daoSet.js');
const m_bo_boSet = require('bo/boSet.js');
const server_DefaultController = require('server/DefaultController.js');
const m_app_invoice_invoiceDao = require('app/invoice/invoiceDao.js');
const m_app_invoice_invoiceBo = require('app/invoice/invoiceBo.js');

function prepare(request, response, responseCallback) {
    var invoiceDao = m_app_invoice_invoiceDao.invoiceDao(request.db);
    var daoSet = m_dao_daoSet.daoSet(request.db, m_app_invoice_invoiceDao.invoiceDao);
    var boSet = m_bo_boSet.boSet(request.db, daoSet, m_app_invoice_invoiceBo.invoiceBo);
    return server_DefaultController.prepare(boSet, request, response, responseCallback);
}

function edit(request, response, responseCallback) {
    var invoiceBo = m_invoiceBo.invoiceBo(request.db);
    return server_DefaultController.edit(invoiceBo, request, response, responseCallback);
}

function list(request, response, responseCallback) {
    var invoiceDao = m_app_invoice_invoiceDao.invoiceDao(request.db);
    var daoSet = m_dao_daoSet.daoSet(request.db, m_app_invoice_invoiceDao.invoiceDao);
    var boSet = m_bo_boSet.boSet(request.db, daoSet, m_app_invoice_invoiceBo.invoiceBo);
    return server_DefaultController.list(boSet, request, response, responseCallback);
}

function templateRow(request, response, responseCallback) {
    var invoiceDao = m_app_invoice_invoiceDao.invoiceDao(request.db);
    var daoSet = m_dao_daoSet.daoSet(request.db, m_app_invoice_invoiceDao.invoiceDao);
    var boSet = m_bo_boSet.boSet(request.db, daoSet, m_app_invoice_invoiceBo.invoiceBo);
    return server_DefaultController.templateRow(boSet, request, response, responseCallback);
}

function saveRow(request, response, responseCallback) {
    var invoiceBo = m_invoiceBo.invoiceBo(request.db);
    return server_DefaultController.saveRow(invoiceBo, request, response, responseCallback);
}

function saveField(request, response, responseCallback) {
    var invoiceBo = m_invoiceBo.invoiceBo(request.db);
    return server_DefaultController.saveField(invoiceBo, request, response, responseCallback);
}

function count(request, response, responseCallback) {
    var invoiceDao = m_app_invoice_invoiceDao.invoiceDao(request.db);
    var daoSet = m_dao_daoSet.daoSet(request.db, m_app_invoice_invoiceDao.invoiceDao);
    var boSet = m_bo_boSet.boSet(request.db, daoSet, m_app_invoice_invoiceBo.invoiceBo);
    return server_DefaultController.count(boSet, request, response, responseCallback);
}

exports.prepare = prepare;
exports.edit = edit;
exports.list = list;
exports.saveField = saveField;
exports.saveRow = saveRow;
exports.count = count;
exports.templateRow = templateRow;
