// server/sql/query.js
/*
 * insert into t ( fields... ) values ( values... )
 * update t set f=v, f=v where f=v
 * delete from t where f=v
 *
 * select f, f from t where f=v
 *
 */

var bjoo = require('BJOObject.js');

var Type = { select: 'select', update: 'update', 'delete': 'delete', insert: 'insert' };

function Query() {
    this._className = 'sql.Query';
    this._type = undefined;
    this._fields = []; // actually, field, or value, or aggregate...
    this._tables = [];
    this._conditions = [];
}

Query.prototype = new bjoo.BJOObject();

Query.prototype.type = function() {
    return this._type;
};
Query.prototype.tables = function(tables) {
    if ( typeof(tables) !== 'undefined' ) {
        this._tables = tables;
        return this;
    }
    return this._tables;
};

Query.prototype.fields = function(fields) {
    if ( typeof(fields) !== 'undefined' ) {
        this._fields = fields;
        return this;
    }
    return this._fields;
};

Query.prototype.conditions = function() {
    if ( typeof(conditions) !== 'undefined' ) {
        this._conditions = conditions;
        return this;
    }
    return this._conditions;
};

Query.prototype.groupBys = function(groupBys) {
    if ( typeof(groupBys) !== 'undefined' ) {
        this._groupBys = groupBys;
        return this;
    }
    return this._groupBys;
};

Query.prototype.groupBy = function(groupBy) {
    this._groupBys.push(groupBy);
    return this;
};

Query.prototype.select = function() {
    this._type = Type.select;
    this._fields = Array.prototype.slice.call(arguments);
    return this;
};

Query.prototype.update = function() {
    this._type = Type.update;
    this._fields = Array.prototype.slice.call(arguments);
    return this;
};

Query.prototype.insert = function() {
    this._type = Type.insert;
    this._fields = Array.prototype.slice.call(arguments);
    return this;
};

Query.prototype.delete = function() {
    this._type = Type.delete;
    return this;
};

Query.prototype.from = function(tables) {
    this._tables = Array.prototype.slice.call(arguments);
    return this;
};

Query.prototype.into = function(table) {
    this._tables = [ table ];
    return this;
};

Query.prototype.table = function(table) {
    if ( typeof(table) !== 'undefined' ) {
        this._tables = [ table ];
        return this;
    }
    return this._tables[0];
};

/**
 * where(condition, condition...)
 */
Query.prototype.where = function() {
    this._conditions = Array.prototype.slice.call(arguments);
    return this;
};

function select() {
    return Query.prototype.select.apply(new Query(), Array.prototype.slice.call(arguments));
}

function insert() {
    return Query.prototype.insert.apply(new Query(), Array.prototype.slice.call(arguments));
}

function update() {
    return Query.prototype.update.apply(new Query(), Array.prototype.slice.call(arguments));
}

function remove() {
    return Query.prototype.delete.apply(new Query(), Array.prototype.slice.call(arguments));
}

exports.Type = Type;
exports.Query = Query;
exports.select = select;
exports.insert = insert;
exports.update = update;
exports.delete = remove;
